src/Avalonia.Base/Utilities/Ref.cs

```diff
@@ -0,0 +1,143 @@
+﻿using System;
+using System.Runtime.ConstrainedExecution;
+using System.Threading;
+
+namespace Avalonia.Utilities
+{
+    public interface IRef<out T> : IDisposable where T : class
+    {
+        T Item { get; }
+        IRef<T> Clone();
+        IRef<TResult> CloneAs<TResult>() where TResult : class;
+    }
+
+    
+
+    public static class RefCountable
+    {
+        public static IRef<T> Create<T>(T item) where T : class, IDisposable
+        {
+            return new Ref<T>(item, new RefCounter(item));
+        }
+
+        public static IRef<T> CreateUnownedNotClonable<T>(T item) where T : class
+            => new TempRef<T>(item);
+
+        class TempRef<T> : IRef<T> where T : class
+        {
+            public void Dispose()
+            {
+                
+            }
+
+            public TempRef(T item)
+            {
+                Item = item;
+            }
+            
+            public T Item { get; }
+            public IRef<T> Clone() => throw new NotSupportedException();
+
+            public IRef<TResult> CloneAs<TResult>() where TResult : class
+                => throw new NotSupportedException();
+        }
+        
+        class RefCounter
+        {
+            private IDisposable _item;
+            private volatile int _refs;
+            private object _lock = new object();
+
+            public RefCounter(IDisposable item)
+            {
+                _item = item;
+            }
+
+            public void AddRef()
+            {
+                Interlocked.Increment(ref _refs);
+            }
+
+            public void Release()
+            {
+                if (Interlocked.Decrement(ref _refs) == 0)
```

Comment:
If we do a CompareExchange loop here we don't need the lock.

Reply:
Lock is needed to ensure that `Item` won't return a value while `Dispose` is called on another thread.

Reply:
I was afraid of AddRef called after the last Release. That should not happen in normal circumstances, so we should be able to remove that lock.

Reply:
The lock object used here is unique to the RefCounter instance and is only used in this method. It won't stop it from returning a value from Item since that's under a different lock

Reply:
We can switch out the `Interlocked.Increment` in `AddRef` for a `CompareExchange` loop that checks for a ref-count of `0` and throws an `ObjectDisposedException` or something if it sees a ref-count of zero when trying to add a reference.